tensorflow/python/keras/layers/wrappers.py

```diff
@@ -148,56 +150,73 @@ def _get_shape_tuple(self, init_tuple, tensor, start_idx, int_shape=None):
+      replaced by the corresponding dimension from `tf.shape(tensor)`.
     """
     # replace all None in int_shape by K.shape
     if int_shape is None:
       int_shape = K.int_shape(tensor)[start_idx:]
     if not any(not s for s in int_shape):
-      return init_tuple + tuple(int_shape)
+      return convert_to_tensor(init_tuple + tuple(int_shape))
     shape = K.shape(tensor)
     int_shape = list(int_shape)
     for i, s in enumerate(int_shape):
       if not s:
         int_shape[i] = shape[start_idx + i]
-    return init_tuple + tuple(int_shape)
+    return convert_to_tensor(init_tuple + tuple(int_shape))
 
   def build(self, input_shape):
-    input_shape = tensor_shape.TensorShape(input_shape).as_list()
-    if len(input_shape) < 3:
+    input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+    input_dims = generic_utils.to_list(nest.map_structure(lambda x: x.ndims, input_shape))
+    if all(dim < 3 for dim in input_dims):
       raise ValueError(
           '`TimeDistributed` Layer should be passed an `input_shape ` '
           'with at least 3 dimensions, received: ' + str(input_shape))
     # Don't enforce the batch or time dimension.
-    self.input_spec = InputSpec(shape=[None, None] + input_shape[2:])
-    child_input_shape = [input_shape[0]] + input_shape[2:]
+    self.input_spec = nest.map_structure(lambda x: InputSpec(shape=[None, None] + x.as_list()[2:]),
+                                         input_shape)
+    child_input_shape = nest.map_structure(
+        lambda x: tensor_shape.TensorShape([x.as_list()[0]] + x.as_list()[2:]),
+        input_shape)
+    child_input_shape = tf_utils.convert_shapes(child_input_shape)
     super(TimeDistributed, self).build(tuple(child_input_shape))
     self.built = True
 
   def compute_output_shape(self, input_shape):
-    input_shape = tensor_shape.TensorShape(input_shape).as_list()
-    child_input_shape = tensor_shape.TensorShape([input_shape[0]] +
-                                                 input_shape[2:])
+    input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+    def remove_timesteps(dims):
+      dims = dims.as_list()
+      return tensor_shape.TensorShape([dims[0]] + dims[2:])
+
+    child_input_shape = nest.map_structure(remove_timesteps, input_shape)
     child_output_shape = self.layer.compute_output_shape(child_input_shape)
-    if not isinstance(child_output_shape, tensor_shape.TensorShape):
-      child_output_shape = tensor_shape.TensorShape(child_output_shape)
-    child_output_shape = child_output_shape.as_list()
-    timesteps = input_shape[1]
-    return tensor_shape.TensorShape([child_output_shape[0], timesteps] +
-                                    child_output_shape[1:])
+    child_output_shape = tf_utils.convert_shapes(child_output_shape,
+                                                 to_tuples=False)
+    timesteps = tf_utils.convert_shapes(input_shape)
+    timesteps = nest.flatten(timesteps)[1]
```

Comment:
This doesn't feel right, since timesteps is now a structure, getting the [1] from the flattened list will give u a shape, not a dim.

Reply:
Since the input shapes are first converted to a nested structure of shape tuples `timesteps = tf_utils.convert_shapes(input_shape)` with the default option `to_tuples=True` (see [here](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/utils/tf_utils.py#L193)), the subsequent `nest.flatten(timesteps)` command will flatten the nested structure of tuples to a list of shape values. Therefore, `nest.flatten(timesteps)[1]` will return the value of the time dimension of the first input tensor.

However, to make it more clear this could be changed to:
```suggestion
    timesteps = tf_utils.convert_shapes(input_shape, to_tuples=True)
    timesteps = nest.flatten(timesteps)[1]
```